### ISSUE TITLE
Respect upper bounds of Version Range when updating in UI, PMC and dotnet package update

### ISSUE THREAD

**Issue by zhili1208** [NuGet/Home#6566]
Hi everyone at NuGet.org.

I’m using new format of the nuget package management that is fully integrated with MSBuild, it is very awesome and work very good except that it have one really odd error.
In old format(packages.config), we could use `allowedVersions` to limit version of allowed nugets, in this new format Version range should specified in `Version` tag of the `PackageReference` but
we have 2 problems with this:
1)	PackagaeManagerConsole does not respect this version limit,
So if I have `Version=”[1.0.0, 1.5.0)”` and there is a version `2.0.0` of the package, UI respect my limitation and show no update but `Update-Package` command will update my
reference to version `2.0.0`
2)	If package updated to a newer version(a version that is allowed in my range), then new version will be written in place of old version and it will remove my version range limit
So if I have `Version=”[1.0.0, 1.5.0)”` and there is version `1.1.0` and `2.0.0` of the package, and I use nuget UI, it will update my project to version `1.1.0` as expected, but as a
side effect it will change my Version tag to `Version=”1.1.0”` and so remove my version limitation, and now it will show that I have an update for version `2.0.0`.

I want to know how I use this new format and use old good `allowedVersions` or something that have same effect.

Ideas in https://github.com/nuget/home/issues/14767 and dotnet package update equivalent in https://github.com/NuGet/Home/issues/14313

**Comment by MattLavalleeMA** [comment]
Same here. Almost 2 years old for a P1??

**Comment by MagicAndre1981** [comment]
Any update on this issue? @nkolev92 @anangaur 

[We need to restrict](https://github.com/NuGet/Home/issues/6763#issuecomment-634489134) the offered nuget updates similar to _allowedVersions_ in `packages.config`.

**Comment by Indigo744** [comment]
This issue is still open, even though it is really a showstopper when multiple people may update packages using the UI. Every time we need to step in and add back the constraints...

This is madness. PackageReference is supposed to be "the new norm" but clearly lacks of basic usability like this, which is known for almost 3 years... Come on!

**Comment by MagicAndre1981** [comment]
@donnie-msft when will this be fixed? In 16.9? Why don't we get any response from you @nkolev92 @anangaur?

I opened my solution and now get tons of .net 5 updates when I want to do updates from .net core 3.1.9 to 3.1.10

![image](https://user-images.githubusercontent.com/8274816/98796027-35283600-240b-11eb-8a78-8bfdbfb48534.png)

I now used my filemanager to search for 3.1.9 to list all csprojs and next used Search&Replace from notepad3 to update from 3.1.9. to 3.1.10  🤮

**Comment by zivkan** [comment]
Please see https://github.com/NuGet/Home/issues/10264 for some additional feedback from customers. In particular, this comment: https://github.com/NuGet/Home/issues/10264#issuecomment-800253081

**Comment by erdembayar** [comment]
@martinrrm 
Is it related to https://github.com/NuGet/Home/issues/9829?

**Comment by donnie-msft** [comment]
This issue has multiple parts. 
**Package Manager console** - I believe isn't improved yet. @martinrrm, we could consider for next sprint,

**PM UI** - We should validate that #9829 or other work has resolved:
-  >  but as a side effect it will change my Version tag to Version=”1.1.0” and so remove my version limitation
- This comment: https://github.com/NuGet/Home/issues/6566#issuecomment-725321964
- This comment: https://github.com/NuGet/Home/issues/6566#issuecomment-640406577

**Comment by nkolev92** [comment]
I think some of the asks in this issue need customer validation/design. 

Specifically, point 1 in the original issue body.

**Comment by adschm** [comment]
I can "validate" that the current situation is still completely unusable if you want to manually update packages via NuGet UI.

1. If we use the limitations via Version, it either updates packages automagically when a new one arrives (which we don't want) or the constraint gets overwritten after manual update of a package.
2. If we do not constrain, we are bombarded with .NET 6 package builds (which we cannot use since AI in VS 2022 eats up all available memory and thus is unusable for the moment).
3. Updating individual packages manually will take hours, since the "Updates" section will take several seconds for reevaluation after each package update.

I've now updated our packages (again) by replacing all 5.0.12 occurrences with 5.0.15 via Notepad++ "Replace in Files". :-(

Unfortunately, it appears to me that this is flawed by design. Obviously, the removal of the allowedVersions has reduced the options to store the relevant information, and with only one variable "Versions" to store information about both current version and version constraints, it's just not possible to transport as much intention as before.

So, in my opinion choices are only to reintroduce an AllowedVersions property again or live with the current feature reduction. The lack of progress for this subject looks like maintainers have decided for the latter ...

**Comment by Nirmal4G** [comment]
Yeah, What's the point of NuGet PowerShell and PM UI if you make us use Notepad++ or Code to upgrade packages through find and replace? That's just sad! With CPVM, I at least have one place to perform search and replace. Still, there's a use of find and replace!!

**Bring back `AllowedVersions` on CPVM at least.**

**Comment by uhfath** [comment]
So is there any workaround if we need to keep current .NET version to 6.x but still update other packages to their latest versions using UI in VS?

**Comment by MagicAndre1981** [comment]
Yeah, I also use search & replace in notepad3 to update nugets to 6.0.x, what a mess.

**Comment by Bellarmine-Head** [comment]
So this ticket is apparently touted as the answer to these problems:-

![image](https://user-images.githubusercontent.com/3470891/204840948-c4852fea-b1bf-4ef6-a824-9c4ed585628f.png)

1) .NET 7 packages are recommended for a .NET 6 project.  (They will fail to install if there is an attempt to do so.)

2) The invalid .NET 7 packages are obscuring valid .NET 6 package updates, which the project/solution might well need for security.

As I have mentioned in other tickets that have been closed (e.g. because they are dups of this ticket, directly or indirectly): imagine Windows Update behaving in this way!  Imagine you're on Windows 10 and Windows Update starts offering you Windows 11 updates which fail to install, and which hide genuine Windows 10 updates.  The scenario is ridiculous of course, yet that is effectively what is happening wrt NuGet packages for .NET projects/solutions.

My understanding is that the problems described above will go away if this ticket is dealt with.

**Comment by StingyJack** [comment]
This is needed so that we can easily prevent versions of the Moq package that are 4.20 and later from being "whoops" upgraded in a project.

**Comment by ThomasHeijtink** [comment]
Now after almost 6 years, and still no improvements. Why is such a fundamental feature being ignored or procrastinated? You only need the package manager to honour the lowest framework version in a solution. This already covers 90% (if not more) of all use cases. You can always improve this later with other settings etc. But the majority of the people have a problem with having to deal a year long (if not longer) with their package update experience hampered and frustrated because the package manager is lacking such fundamental and basic feature.

**Comment by MagicAndre1981** [comment]
the Microsoft managers simply don't do coding and have no idea what is important and what not.

**Comment by Bellarmine-Head** [comment]
At a glacial pace, something might be happening: https://github.com/NuGet/Home/issues/5725#issuecomment-1810525684

**Comment by mr-davidc** [comment]
Just stumbled across this issue and am kinda shocked that functionality that exists in a `packages.config` file isn't available in the "new" `PackageReference` format...

**Comment by StingyJack** [comment]
> Just stumbled across this issue and am kinda shocked that functionality that exists in a `packages.config` file isn't available in the "new" `PackageReference` format...

@mr-davidc  - They broke a bunch of stuff with the move to package reference.  install.ps1 scripts no longer work, no more xdt support, and one other thing that I dont recall (because I didn't use it).   Not to mention all of the crippling bugs that came with Package referece.

**Comment by MagicAndre1981** [comment]
any update on this issue?

**Comment by ShockwaverReal** [comment]
Are we really stuck on this? I mean, it seems quite fundamental to me...
Plus, the UI for me it's showing the very latest version available regardless of what I specify in `version` attribute.

**Comment by Bellarmine-Head** [comment]
@ShockwaverReal - yes it seems we're really stuck on this.  Another year, another .NET release, same old problem.

![image](https://github.com/user-attachments/assets/9ded3540-34da-429b-96e7-7f8523be633e)

Here NuGet is offering to install an incompatible version of the packages, and *hiding* compatible versions that I might well need to upgrade to for security.  Just like last year, and the year before, and the year before, and etc.  Maybe they can ask "AI" to sort it out...

**Comment by ShockwaverReal** [comment]
@Bellarmine-Head - It's grotesque... in order to update hundreds of references to the latest compatible versions in my projects I have to manually update them **_ONE BY ONE_** through UI or csproj... Insane.

**Comment by mexmer** [comment]
> @Bellarmine-Head - It's grotesque... in order to update hundreds of references to the latest compatible versions in my projects I have to manually update them **_ONE BY ONE_** through UI or csproj... Insane.

it's stupid and it was giving me headache, so what i'm doing is opening csproj, and with search and replace i replace all versions, then load project, and since i have automatic package restore enabled in visual studio, it does indeed download all missing packages.

yes, it's clunky workaround, but works ... it's maintable if you are doing it for handfull of project files, can't image doing it for large number of projects tho' ... maybe writing some automation tool.

**Comment by MagicAndre1981** [comment]
> it's stupid and it was giving me headache, so what i'm doing is opening csproj, and with search and replace i replace all versions, then load project, and since i have automatic package restore enabled in visual studio, it does indeed download all missing packages.

this is what I also do for [4 years now](https://github.com/NuGet/Home/issues/6566#issuecomment-725321964)

**Comment by MikeFarrington** [comment]
It's not perfect, but this command is helpful in identifying which packages have minor updates.

```
dotnet list package --outdated --highest-minor --highest-patch
```

**Comment by PeterKnaggs** [comment]
Six years now.

This is still and issue in 17.12.3

**Comment by StingyJack** [comment]
No programmer should be updating files ONE by ONE.  Use a tool like notepad++ to do find and replace in files. If that doesn't work for you, then make a tool or script. 

[Powershell is great for dealing with XML and proj files.](https://github.com/StingyJack/Vs-Utility)

**Comment by dominikjeske** [comment]
I just created powershell script for update nugets with ommiting non LTS versions. I think that with LTS release schedule this is standard behavior for many people out there. This feature is a MUST for me.

**Comment by vernou** [comment]
@dominikjeske, nice. Can you share the script?

**Comment by dominikjeske** [comment]
Just for a note - my script only lists dependencies because I want control what I'm updating but It is good starting point if you want also update

``` powershell
Write-Host "Reading dependencies for all versions..." -ForegroundColor Gray
$majorDepRaw = dotnet list package --outdated -f "net8.0" --format json

Write-Host "Reading dependencies for minior versions..." -ForegroundColor Gray
$miniorDepRaw = dotnet list package --outdated -f "net8.0" --highest-minor --format json

$majorDep = $majorDepRaw | ConvertFrom-Json
$miniorDep = $miniorDepRaw  | ConvertFrom-Json

$miniorBlackList = @("Microsoft.Extensions", "Microsoft.AspNetCore")
function ParseDependencies()
{
    [CmdletBinding()]
    param (
        $dependencies
    )
    $packages = @{}

    $dependencies.projects | ForEach-Object {
        $project = $_
        $project.frameworks | Where-Object { $_ -ne $null } | ForEach-Object {
            $framework = $_
            $framework.topLevelPackages | ForEach-Object {
                $package = $_
                if(!$packages.ContainsKey($package.id)) {
                    $packages[$package.id] = [PSCustomObject]@{
                        Id = $package.id
                        CurrentVersion = $package.resolvedVersion
                        NewVersion = $package.latestVersion
                    }
                }
            }
        }
    } 
    $packages
}

$packages = ParseDependencies $majorDep

Write-Host "---------------------All---------------------" -ForegroundColor Green
$packages.Values | Sort-Object -Property Id | ForEach-Object {
    $package = $_
    $skip = $false
    $miniorBlackList | ForEach-Object {
        $miniorBlack = $_
        if($package.Id.StartsWith($miniorBlack)) {
            $skip = $true
        }
    }
    if(!$skip) {
        Write-Host "$($package.Id) $($package.CurrentVersion)->$($package.NewVersion)"
    }
}

Write-Host "---------------------Minior---------------------" -ForegroundColor Green
$packagesMinior = ParseDependencies $miniorDep

$packagesMinior.Values | Sort-Object -Property Id | ForEach-Object {
    $package = $_
    $skip = $true
    $miniorBlackList | ForEach-Object {
        $miniorBlack = $_
        if($package.Id.StartsWith($miniorBlack)) {
            $skip = $false
        }
    }
    if(!$skip) {
        Write-Host "$($package.Id) $($package.CurrentVersion)->$($package.NewVersion)"
    }
}

```

**Comment by daTooz** [comment]
Bump +1

**Comment by voroninp** [comment]
After FluentAssertion case I've discovered that if I specify a range as `Version="[7.*, 8)"` Package Management Console won't even show me the availability of v8.

**Comment by jbennink** [comment]
I added a lot of testing with PMC and version ranges to this issue #14072 since I did not find this **much older** issue first. I don't want to copy paste those comments but as of 2026 this is still going on. Also @voroninp remark is correct, but I have seen that setting a limit like [8.5.2, 9.0.0) would still show me a 9.0.1 update.

**Comment by Saibamen** [comment]
https://github.com/NuGet/NuGet.Client/pull/7323
